### ISSUE TITLE
New format : Bill of Materials

### ISSUE THREAD

**Issue by Cywil23** [SemanticMediaWiki/SemanticResultFormats#204]
Hello,

This is not really an issue but a proposition for a new format to make a Bill of materials.

We are trying to use the tree format, but we have several instances of our items.
It's exactly the same case that the example used [here](https://www.semantic-mediawiki.org/wiki/Help:Tree_format#Example)
England appears two times (first in top level, and second under United Kingdom). But the "children" appear only at the first time.
Is it possible to make appear the children of England (Greenwich and London) for the twices?
Our example is [here](http://wiki.opensourceecologie.org/index.php/Nomenclature)
And we would like to show B-1-1 as the child of the twice B-2-0 (and not only for the first)

Then, we would like to have a shaping like [this file ](https://docs.google.com/spreadsheets/d/12aApzR5RbvaR_GyY5O9SLabGVV8ikGLzfGQNb_YDx80/edit#gid=355111870) with the first column is the level of the item

Thanks a lot by advanced for your answer

**Comment by kghbln** [comment]
Thanks for your question and request. I do not think that it is an good idea to create a result format to solve a wiki specific problem but rather have result format that could serve a lot of wikis.

Since you are referring to the tree format probably @s7eph4n can provide some answers as well as an insight as for the feasibility.

**Comment by s7eph4n** [comment]
Not allowing child elements for the second appearance of a parent element was done to avoid infinite recursion. The implemented logic is rather simple and can be found in https://github.com/SemanticMediaWiki/SemanticResultFormats/blob/master/formats/tree/SRF_Tree.php#L70

To implement a more sophisticated circle detection would not really be a problem (just walk up the tree until you either find a root element (=> good) or an element already visited (=> circle!)), but it would be a more complex algorithm, that for larger datasets might become a performance bottleneck. Pull requests welcome.

To build a quasi tree format, where you would show the level of an element in an extra column is certainly doable, but it would be a rather specialised result format, so I agree with Karsten that you will probably not find somebody to just do it. IMO your best bet would be to either develop it yourself or <a href="#" title="pay">motivate</a> somebody to do it for you.

**Comment by b3j0f** [comment]
A tree with circles ?! It sounds so weird...

Anyway, I would like to know if adding a "maxdepth" parameter is a simpler and quicker solution for you ?

**Comment by s7eph4n** [comment]
Well, SMW is not limited to storing trees. You can store any graph, so circles are a possibility that the tree format has to take into account. A maxdepth could indeed help with that. But I think the first question really is to find a good algorithm to develop the tree from the elements. If it involves a maxdepth parameter, go for it. If not, even better.

Fun question: Consider a structure like this:
```
    ----------
   |          |
   v         bar
  foo         ^
   |          |
   v        quok
  baz         ^
   |          |
    ----------
```
Which one should be the root element?

**Comment by b3j0f** [comment]
"foo" of course ! ;)

I thought the tree format require to choose a root before beeing displayed.

I use to handle a maxdepth parameter in such cases, for memory reasons if tree elements are references.

Well, I will see next week if I can propose a pull request easily.

Thanks for your quick responses :)

**Comment by Cywil23** [comment]
Hi @kghbln 

> I do not think that it is an good idea to create a result format to solve a wiki specific problem but rather have result format that could serve a lot of wikis.

This format could serve all wikis about [Product Data Management (PDM)](https://en.wikipedia.org/wiki/Product_data_management), so about all [Wikifab](https://github.com/Wikifab) wikis and all Open Source Hardware wikis.

Thansk @s7eph4n for your answers.

If somebody (@ClementFlipo , @pierreboutet, @Padhh ?) would like to help us on this topic, please find our first commit on my fork [here](https://github.com/Cywil23/SemanticResultFormats/commit/5a16bb7e4dfc2b438f44acabf696b19ecabdb2b0) (thanks to @b3j0f )

**Comment by alex-mashin** [comment]
Fixed by #226, isn't it?

**Comment by krabina** [comment]
Seems to be fixed, thus closing. Feel free to reopen.
